### PR TITLE
[SPARK-46173][SQL] Skipping trimAll call during date parsing

### PR DESCRIPTION
**Pull request by dbatomic**
### What changes were proposed in this pull request?

This PR is a response to a user complaint that stated that date to string casts are too slow if input string has many whitespace/isocontrol character as it's prefix or sufix.

Current behaviour is that in StringToDate function we first call trimAll function to remove any whitespace/isocontrol chars. TrimAll creates new string and then we work against that string. This is not really needed since StringToDate can just simply skip these characters and do all the processing in single loop without extra allocations. Proposed fix is exactly that.

### Why are the changes needed?

These changes should drastically improve edge case where input string in cast to date has many whitespace as prefix/sufix.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Extending existing tests to cover both prefix and suffix cases with whitespaces/control chars.

The change also includes a small unit benchmark for this particular case. Looking for feedback if we want to keep the benchmark, given that this is a rather esoteric edge case.

The benchmark measures time needed to do 10k calls of stringToDate function against a input strings that are in format <65k whitespace><correct_date><65 whitespace> and <correct_date><65k whitespace>. Here are the results:

| input example | prior to change | after the change |
| -------------- | ------------------ | ------------- |
| 65k whitespace prefix + 65k suffix | 3250ms | 484ms |
| 65k whitespace suffix | 1572ms | <1ms |

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dbatomic** [comment]
@MaxGekk -> Looking for your feedback on this one.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 1]
```diff
@@ -305,21 +305,28 @@ trait SparkDateTimeUtils {
...
+    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) {
+      j += 1;
+    }
+
+    if (j < bytes.length && (bytes(j) == '-' || bytes(j) == '+')) {
       sign = if (bytes(j) == '-') -1 else 1
       j += 1
     }
-    while (j < bytes.length && (i < 3 && !(bytes(j) == ' ' || bytes(j) == 'T'))) {
+    while ( j < bytes.length &&
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:328]: Could you remove the gap and leave this part as it was:
```suggestion
    while (j < bytes.length &&
```

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 2]
```diff
@@ -305,21 +305,28 @@ trait SparkDateTimeUtils {
       (segment == 0 && digits >= 4 && digits <= maxDigitsYear) ||
         (segment != 0 && digits > 0 && digits <= 2)
     }
-    if (s == null || s.trimAll().numBytes() == 0) {
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:308]: So, actually we could still have the optimization if you add an if after:
```scala
    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) {
      j += 1;
    }
    if (j == bytes.length) {
      return None;
    }
```

**Comment by MaxGekk** [comment]
@dbatomic Could you remove this Link to [JIRA](https://issues.apache.org/jira/browse/SPARK-46173) from PR's description (PR's title has it already), and add the `[SQL]` tag to the title.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 3]
```diff
@@ -305,21 +305,28 @@ trait SparkDateTimeUtils {
...
+      j += 1;
+    }
+
+    if (j < bytes.length && (bytes(j) == '-' || bytes(j) == '+')) {
       sign = if (bytes(j) == '-') -1 else 1
       j += 1
     }
-    while (j < bytes.length && (i < 3 && !(bytes(j) == ' ' || bytes(j) == 'T'))) {
+    while ( j < bytes.length &&
+      (i < 3 && !(UTF8String.isWhitespaceOrISOControl(bytes(j)) || bytes(j) == 'T'))) {
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:329]: It would be nice to exclude the check `UTF8String.isWhitespaceOrISOControl` from the main loop. In regular cases, dates don't have any spaces, so, don't need to check every char - isn't a whitespace or not. Can we just inline `trimAll` at the beginning? like:
```scala
    var j = 0

    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) j += 1
    var end = bytes.length - 1
    while (j < end && UTF8String.isWhitespaceOrISOControl(bytes(end))) end -= 1
    if (j >= end) return None

    if (bytes(j) == '-' || bytes(j) == '+') {
      sign = if (bytes(j) == '-') -1 else 1
      j += 1
    }
    while (j < end && (i < 3 && !(bytes(j) == 'T'))) {
```

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 4]
```diff
@@ -502,7 +502,7 @@ private UTF8String copyUTF8String(int start, int end) {
    * Determines if the specified character (Unicode code point) is white space or an ISO control
    * character according to Java.
    */
-  private boolean isWhitespaceOrISOControl(int codePoint) {
+  public static boolean isWhitespaceOrISOControl(int codePoint) {
```
Comment (beliefer) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:505]: Because the visibility changed, please move `isWhitespaceOrISOControl` above the constructor.

**Comment by MaxGekk** [comment]
> The change also includes a small unit benchmark for this particular case.

I wonder of other benchmarks. Do you observe perf regressions? I am asking just in case.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:329]: ok, I see. Let's preserve the behaviour.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:328]: Just in case, here is the style guide: https://github.com/databricks/scala-style-guide

**Comment by dbatomic** [comment]
> > The change also includes a small unit benchmark for this particular case.
> 
> I wonder of other benchmarks. Do you observe perf regressions? I am asking just in case.

I just ran Parsing benchmarks locally. Here are the results:

With trimming changes:
func_name:                                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
to timestamp str wholestage on                       85             87           2         11.7          85.2       1.1X
to_timestamp wholestage on                          421            423           3          2.4         420.7       1.0X
to_unix_timestamp wholestage on                     418            423           4          2.4         417.9       1.0X
to date str wholestage on                            98            101           1         10.2          98.3       0.9X
to_date wholestage on                               612            616           4          1.6         611.8       1.0X
to_date_with_trimming                               758            763           7          0.0    75792262.5       1.0X
to_date_with_trimming_only_suffix                     0              1           0          0.0       44237.5    1713.3X

--
Without my change
func_name:                                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
to timestamp str wholestage on                       82             83           1         12.2          82.1       1.2X
to_timestamp wholestage on                          418            421           2          2.4         418.1       0.9X
to_unix_timestamp wholestage on                     414            416           2          2.4         413.9       0.9X
to date str wholestage on                           101            103           2          9.9         101.4       0.9X
to_date wholestage on                               620            624           4          1.6         620.3       1.0X

TLDR - no observable change. One some functions it is <1% worse and somewhere it is 1% better, but again, these are just tests on my local machine, without any real isolation so tiny diff is expected.

**Review by MaxGekk — COMMENTED** [review]
LGTM, waiting for CI and either updated benchmark results or exclude them.

**Comment by dbatomic** [comment]
> > Looking for feedback if we want to keep the benchmark, given that this is a rather esoteric edge case.
> 
> I am ok to exclude the benchmark for the pretty specific case from the PR.

I decided to exclude the benchmark - thinking is that there is no point in having a benchmark for such an edge case.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 5]
```diff
@@ -305,21 +305,32 @@ trait SparkDateTimeUtils {
...
+    if (s == null) {
       return None
     }
+
     val segments: Array[Int] = Array[Int](1, 1, 1)
     var sign = 1
     var i = 0
     var currentSegmentValue = 0
     var currentSegmentDigits = 0
-    val bytes = s.trimAll().getBytes
```
Comment (beliefer) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:316]: We can reuse the `trimLeft` here.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 6]
```diff
@@ -305,21 +305,32 @@ trait SparkDateTimeUtils {
...
+    val bytes = s.getBytes
     var j = 0
+
+    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) {
+      j += 1;
+    }
```
Comment (beliefer) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:322]: After using `trimLeft`, we can delete the three lines.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:316]: According to the PR description, we don't want to use these trim methods to avoid string copy.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 7]
```diff
@@ -305,21 +305,32 @@ trait SparkDateTimeUtils {
...
+      return None;
+    }
+
     if (bytes(j) == '-' || bytes(j) == '+') {
       sign = if (bytes(j) == '-') -1 else 1
       j += 1
     }
-    while (j < bytes.length && (i < 3 && !(bytes(j) == ' ' || bytes(j) == 'T'))) {
+    while (j < bytes.length &&
+      (i < 3 && !(UTF8String.isWhitespaceOrISOControl(bytes(j)) || bytes(j) == 'T'))) {
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:333]: let's add a comment to explain the behavior: spaces in the middle may be allowed.

Comment (cloud-fan) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:333]: oh is this a behavior change? now not only spaces, but also ISO controls are allowed in the middle.

Comment (cloud-fan) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:333]: Maybe we should calculate the end offset (not always `bytes.length`) at the beginning, by excluding white spaces and ISO controls from the end.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 5 above)
Comment (beliefer) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:316]: Got it.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 6 above)
Comment (beliefer) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:322]: Please ignore the above comment.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 7 above)
Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala]: Implemented the approach with end offset. Please take a look.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 8]
```diff
@@ -305,21 +305,41 @@ trait SparkDateTimeUtils {
...
+    }
+
+    var strlen = bytes.length
+    while (strlen > j && UTF8String.isWhitespaceOrISOControl(bytes(strlen - 1))) {
+      strlen -= 1;
+    }
+
+    if (j >= strlen)
+    {
+      return None
```
Comment (beliefer) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:335]: Is this path reachable?

Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala]: good catch, thanks! Removed.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 9]
```diff
@@ -305,21 +305,35 @@ trait SparkDateTimeUtils {
...
+
     val segments: Array[Int] = Array[Int](1, 1, 1)
     var sign = 1
     var i = 0
     var currentSegmentValue = 0
     var currentSegmentDigits = 0
-    val bytes = s.trimAll().getBytes
+    val bytes = s.getBytes
     var j = 0
+    var strEndTrimmed = bytes.length
```
Comment (beliefer) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:319]: I like the previous name, it was more universal. e.g. `strLen` or `len`.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 10]
```diff
@@ -305,21 +305,35 @@ trait SparkDateTimeUtils {
...
+
+    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) {
```
Comment (beliefer) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:321]: Why not reuse `strEndTrimmed` here?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 11]
```diff
@@ -305,21 +305,35 @@ trait SparkDateTimeUtils {
...
+      j += 1;
+    }
+    if (j == bytes.length) {
```
Comment (beliefer) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:324]: ditto.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 9 above)
Comment (beliefer) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:319]: How about
```
val strLen = bytes.length
var strEndTrimmed = strLen
```
We keep `strLen` immutable and reuse it later. `strEndTrimmed` used as mutable variable.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 10 above)
Comment (cloud-fan) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:321]: how?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 9 above)
Comment (cloud-fan) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:319]: `bytes.length` is very cheap and clear. It doesn't seem helpful to create a variable for it.

Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:319]: Yeah, I also don't think that we need two variables. I can call it strLen or strLenTrimmed or anything along those lines, if that sounds better to you guys.

Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:319]: Btw, we don't need strLen/bytes.length at all after we get strEndTrimmed var. We don't care about anything after strEndTrimmed in this function.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 10 above)
Comment (beliefer) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:321]: Please see
https://github.com/apache/spark/pull/44110#discussion_r1416537457
What I mean is reuse `strLen`

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 9 above)
Comment (beliefer) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:319]: I'm OK keep it unchanged here too.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala (hunk 10 above)
Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:321]: so, you are basically suggesting to rename strEndTrimmed back to strLen?

Comment (cloud-fan) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:321]: `strEndTrimmed` is more accurate.

**Comment by cloud-fan** [comment]
The docker integration test failure is unrelated. merging to master, thanks!

**Comment by MaxGekk** [comment]
@dbatomic Does this suffer from the same issue?
https://github.com/apache/spark/blob/b4d90dd2a622beda542a7ce1ee15af9f312f9724/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L429

If so, could you optimize it too.

**Comment by dbatomic** [comment]
> @dbatomic Does this suffer from the same issue?
> 
> https://github.com/apache/spark/blob/b4d90dd2a622beda542a7ce1ee15af9f312f9724/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L429
> 
> If so, could you optimize it too.

Sure, good catch. I will follow up with PR that takes care of this one.
